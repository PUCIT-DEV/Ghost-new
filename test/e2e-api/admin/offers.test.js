const {agentProvider, mockManager, fixtureManager, matchers} = require('../../utils/e2e-framework');
const {anyEtag, anyObjectId, anyUuid, anyISODateTime, anyISODate, anyString, anyArray, anyLocationFor, anyErrorId} = matchers;

const assert = require('assert');
const nock = require('nock');
const should = require('should');
const sinon = require('sinon');
const testUtils = require('../../utils');
const Papa = require('papaparse');

const models = require('../../../core/server/models');
const { any } = require('bluebird');

let agent;

async function getPaidProduct() {
    return await models.Product.findOne({type: 'paid'});
}

async function getFreeProduct() {
    return await models.Product.findOne({type: 'free'});
}

describe('Offers API', function () {
    let defaultTier;
    let savedOffer;

    before(async function () {
        agent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init();
        await agent.loginAsOwner();
        defaultTier = await getPaidProduct();
    });

    beforeEach(function () {
        mockManager.mockMail();
    });

    afterEach(function () {
        mockManager.restore();
    });

    it('Has no initial offers', async function () {
        await agent
            .get(`offers/`)
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag
            })
            .matchBodySnapshot({
                offers: []
            });
    });

    it('Can add a new offer', async function () {
        const newOffer = {
            name: 'Black Friday',
            code: 'black-friday',
            display_title: 'Black Friday Sale!',
            display_description: '10% off on yearly plan',
            type: 'percent',
            cadence: 'year',
            amount: 12,
            duration: 'once',
            duration_in_months: null,
            currency_restriction: false,
            currency: null,
            status: 'active',
            redemption_count: 0,
            tier: {
                id: defaultTier.id
            }
        };
        
        const {body} = await agent
            .post(`offers/`)
            .body({offers: [newOffer]})
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag,
                location: anyLocationFor('offers')
            })
            .matchBodySnapshot({
                offers: [{
                    id: anyObjectId,
                    tier: {
                        id: anyObjectId
                    }
                }]
            });
        savedOffer = body.offers[0];
    });

    it('Can browse', async function () {
        await agent
            .get(`offers/`)
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag
            })
            .matchBodySnapshot({
                offers: new Array(1).fill({
                    id: anyObjectId,
                    tier: {
                        id: anyObjectId
                    }
                })
            });
    });

    /** 
     * @todo: Can browse with filter
     */

    it('Can get a single offer', async function () {
        await agent
            .get(`offers/${savedOffer.id}/`)
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag
            })
            .matchBodySnapshot({
                offers: new Array(1).fill({
                    id: anyObjectId,
                    tier: {
                        id: anyObjectId
                    }
                })
            });
    });

    it('Can edit an offer', async function () {
        // We can change all fields except discount related fields
        let updatedOffer = {
            name: 'Cyber Monday',
            code: 'cyber-monday',
            display_title: 'Cyber Monday Sale!',
            display_description: '10% off on yearly plan, only today',
        };

        await agent
            .put(`offers/${savedOffer.id}`)
            .body({offers: [updatedOffer]})
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag
            })
            .matchBodySnapshot({
                offers: new Array(1).fill({
                    id: anyObjectId,
                    tier: {
                        id: anyObjectId
                    }
                })
            })
            .expect(({body}) => {
                body.offers[0].should.match(updatedOffer);
            });
    });

    it('Can archive an offer', async function () {
        // We can change all fields except discount related fields
        let updatedOffer = {
            status: 'archived'
        };

        await agent
            .put(`offers/${savedOffer.id}`)
            .body({offers: [updatedOffer]})
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag
            })
            .matchBodySnapshot({
                offers: new Array(1).fill({
                    id: anyObjectId,
                    tier: {
                        id: anyObjectId
                    }
                })
            })
            .expect(({body}) => {
                body.offers[0].should.match(updatedOffer);
            });
    });

    it('Cannot update offer cadence', async function () {
        // We can change all fields except discount related fields
        let updatedOffer = {
            cadence: 'month'
        };

        await agent
            .put(`offers/${savedOffer.id}`)
            .body({offers: [updatedOffer]})
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag
            })
            .matchBodySnapshot({
                offers: new Array(1).fill({
                    id: anyObjectId,
                    tier: {
                        id: anyObjectId
                    }
                })
            })
            .expect(({body}) => {
                body.offers[0].cadence.should.eql('year');
            });
    });

    it('Cannot update offer amount', async function () {
        // We can change all fields except discount related fields
        let updatedOffer = {
            amount: 20
        };

        // No validation errors are thrown at the moment
        await agent
            .put(`offers/${savedOffer.id}`)
            .body({offers: [updatedOffer]})
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag
            })
            .matchBodySnapshot({
                offers: new Array(1).fill({
                    id: anyObjectId,
                    tier: {
                        id: anyObjectId
                    }
                })
            })
            .expect(({body}) => {
                body.offers[0].amount.should.eql(12);
            });
    });

    it('Cannot update offer tier', async function () {
        // We can change all fields except discount related fields
        const freeTier = await getFreeProduct();
        let updatedOffer = {
            tier: {
                id: freeTier.id
            }
        };

        // No validation errors are thrown at the moment
        await agent
            .put(`offers/${savedOffer.id}`)
            .body({offers: [updatedOffer]})
            .expectStatus(200)
            .matchHeaderSnapshot({
                etag: anyEtag
            })
            .matchBodySnapshot({
                offers: new Array(1).fill({
                    id: anyObjectId,
                    tier: {
                        id: anyObjectId
                    }
                })
            })
            .expect(({body}) => {
                body.offers[0].tier.id.should.eql(defaultTier.id);
            });
    });
});