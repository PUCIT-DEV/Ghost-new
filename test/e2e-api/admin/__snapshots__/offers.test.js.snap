// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offers API Can add a new offer 1: [body] 1`] = `
Object {
  "offers": Array [
    Object {
      "amount": 12,
      "cadence": "year",
      "code": "black-friday",
      "currency": null,
      "currency_restriction": false,
      "display_description": "10% off on yearly plan",
      "display_title": "Black Friday Sale!",
      "duration": "once",
      "duration_in_months": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Black Friday",
      "redemption_count": 0,
      "status": "active",
      "tier": Object {
        "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      },
      "type": "percent",
    },
  ],
}
`;

exports[`Offers API Can add a new offer 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "389",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/offers\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Origin, Accept-Encoding",
  "x-cache-invalidate": "/*",
  "x-powered-by": "Express",
}
`;

exports[`Offers API Can archive an offer 1: [body] 1`] = `
Object {
  "offers": Array [
    Object {
      "amount": 12,
      "cadence": "year",
      "code": "cyber-monday",
      "currency": null,
      "currency_restriction": false,
      "display_description": "10% off on yearly plan, only today",
      "display_title": "Cyber Monday Sale!",
      "duration": "once",
      "duration_in_months": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Cyber Monday",
      "redemption_count": 0,
      "status": "archived",
      "tier": Object {
        "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
        "name": "Default Product",
      },
      "type": "percent",
    },
  ],
}
`;

exports[`Offers API Can archive an offer 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "428",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-cache-invalidate": "/*",
  "x-powered-by": "Express",
}
`;

exports[`Offers API Can browse 1: [body] 1`] = `
Object {
  "offers": Array [
    Object {
      "amount": 12,
      "cadence": "year",
      "code": "black-friday",
      "currency": null,
      "currency_restriction": false,
      "display_description": "10% off on yearly plan",
      "display_title": "Black Friday Sale!",
      "duration": "once",
      "duration_in_months": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Black Friday",
      "redemption_count": 0,
      "status": "active",
      "tier": Object {
        "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
        "name": "Default Product",
      },
      "type": "percent",
    },
  ],
}
`;

exports[`Offers API Can browse 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "414",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Offers API Can edit an offer 1: [body] 1`] = `
Object {
  "offers": Array [
    Object {
      "amount": 12,
      "cadence": "year",
      "code": "cyber-monday",
      "currency": null,
      "currency_restriction": false,
      "display_description": "10% off on yearly plan, only today",
      "display_title": "Cyber Monday Sale!",
      "duration": "once",
      "duration_in_months": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Cyber Monday",
      "redemption_count": 0,
      "status": "active",
      "tier": Object {
        "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
        "name": "Default Product",
      },
      "type": "percent",
    },
  ],
}
`;

exports[`Offers API Can edit an offer 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "426",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-cache-invalidate": "/*",
  "x-powered-by": "Express",
}
`;

exports[`Offers API Can get a single offer 1: [body] 1`] = `
Object {
  "offers": Array [
    Object {
      "amount": 12,
      "cadence": "year",
      "code": "black-friday",
      "currency": null,
      "currency_restriction": false,
      "display_description": "10% off on yearly plan",
      "display_title": "Black Friday Sale!",
      "duration": "once",
      "duration_in_months": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Black Friday",
      "redemption_count": 0,
      "status": "active",
      "tier": Object {
        "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
        "name": "Default Product",
      },
      "type": "percent",
    },
  ],
}
`;

exports[`Offers API Can get a single offer 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "414",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Offers API Cannot update offer amount 1: [body] 1`] = `
Object {
  "offers": Array [
    Object {
      "amount": 12,
      "cadence": "year",
      "code": "cyber-monday",
      "currency": null,
      "currency_restriction": false,
      "display_description": "10% off on yearly plan, only today",
      "display_title": "Cyber Monday Sale!",
      "duration": "once",
      "duration_in_months": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Cyber Monday",
      "redemption_count": 0,
      "status": "archived",
      "tier": Object {
        "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
        "name": "Default Product",
      },
      "type": "percent",
    },
  ],
}
`;

exports[`Offers API Cannot update offer amount 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "428",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-cache-invalidate": "/*",
  "x-powered-by": "Express",
}
`;

exports[`Offers API Cannot update offer cadence 1: [body] 1`] = `
Object {
  "offers": Array [
    Object {
      "amount": 12,
      "cadence": "year",
      "code": "cyber-monday",
      "currency": null,
      "currency_restriction": false,
      "display_description": "10% off on yearly plan, only today",
      "display_title": "Cyber Monday Sale!",
      "duration": "once",
      "duration_in_months": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Cyber Monday",
      "redemption_count": 0,
      "status": "archived",
      "tier": Object {
        "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
        "name": "Default Product",
      },
      "type": "percent",
    },
  ],
}
`;

exports[`Offers API Cannot update offer cadence 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "428",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-cache-invalidate": "/*",
  "x-powered-by": "Express",
}
`;

exports[`Offers API Cannot update offer tier 1: [body] 1`] = `
Object {
  "offers": Array [
    Object {
      "amount": 12,
      "cadence": "year",
      "code": "cyber-monday",
      "currency": null,
      "currency_restriction": false,
      "display_description": "10% off on yearly plan, only today",
      "display_title": "Cyber Monday Sale!",
      "duration": "once",
      "duration_in_months": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Cyber Monday",
      "redemption_count": 0,
      "status": "archived",
      "tier": Object {
        "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
        "name": "Default Product",
      },
      "type": "percent",
    },
  ],
}
`;

exports[`Offers API Cannot update offer tier 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "428",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-cache-invalidate": "/*",
  "x-powered-by": "Express",
}
`;

exports[`Offers API Has no initial offers 1: [body] 1`] = `
Object {
  "offers": Array [],
}
`;

exports[`Offers API Has no initial offers 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "http://127.0.0.1:2369",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "13",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "vary": "Origin, Accept-Encoding",
  "x-powered-by": "Express",
}
`;
